### DISCUSSION TITLE
what causes a "CastedPrimitive"

### DISCUSSION THREAD

**Discussion by sampaioletti** [fzyzcjy/flutter_rust_bridge#2436] [Q&A]
I'm not sure if this is a bug or an issue on my end...but I have some enums that have i64 fields i.e.

```rust
pub enum SomeType{
    Field0{val: i64}
}
```
the generated function for converting that field is

```dart
@protected
int dco_decode_CastedPrimitive_i_64(dynamic raw) {
  // Codec=Dco (DartCObject based), see doc to use other codecs
  throw UnimplementedError(
      'Not implemented in this codec, please use the other one');
}
```

if I manually replace that function with

```dart
@protected
int dco_decode_CastedPrimitive_i_64(dynamic raw) {
  // Codec=Dco (DartCObject based), see doc to use other codecs
  // throw UnimplementedError(
  //     'Not implemented in this codec, please use the other one');
  return dco_decode_i_64(raw);
}
```
my code executes fine I'm using `type_64bit_int: true` in my flutter_rust_bridge.yaml is this a side effect of that? Is there a reason I should have to switch to sse just because of the i64?

i noticed that chrono is using the dco method

```dart
@protected
DateTime dco_decode_Chrono_Utc(dynamic raw) {
  // Codec=Dco (DartCObject based), see doc to use other codecs
  return dcoDecodeTimestamp(ts: dco_decode_i_64(raw).toInt(), isUtc: true);
}
```

**Comment by sampaioletti** [comment]
If I remove the `type_64bit_int: true` from flutter_rust_bridge.yaml it works on native (windows) but then it fails because the generated freeezed code is using int instead of bigint. I'm guessing since the freezed code is generated for the current platform (windows) where the alias is int..but then when compiled for web it is BigInt i've never tried different freezed generation based on the platform. Or if freezed can respect the type alias `PlatformInt64` somehow.

**Reply by sampaioletti** [reply]
Looks like freezed supporting the alias is blocked by
https://github.com/rrousselGit/freezed/issues/628

**Reply by fzyzcjy** [reply]
>  Is there a reason I should have to switch to sse just because of the i64?

It is just a not-yet-implemented thing - DCO/CST codec has not implemented this. Use `full_dep: false` (or leave it unspecified) will use SSE by default.

**Reply by sampaioletti** [reply]
Thanks for the response.

I dont have full_dep set in my yaml but it tries to use dco for both internal and external i64s

```yaml
rust_input: crate::api,ui-types
rust_root: rust/
dart_output: lib/src/rust
type_64bit_int: true
stop_on_error: false
rust_preamble: |
  use crate::frb_helper::*;
```
using latest release for codegen

**Reply by fzyzcjy** [reply]
My personal suggestion is to just use SSE (which is by default) now since DCO/CST has not implemented Casted Primitive, but feel free to PR if you like as well!

**Reply by sampaioletti** [reply]
Thanks I might do a PR seems simple enough...I think I am missing something though... you said SSE is the default, what am I doing that is making it not use the default. I'm not specifying anything anywhere and the docs make it seem like DCO is the default?

Thanks!

**Reply by fzyzcjy** [reply]
Yes that looks confusing... Could you please make a minimal reproducible sample? Probably fork this repo, modify frb_example/dart_minimal as little as possible, to reproduce this.

**Reply by sampaioletti** [reply]
Ok doing that helped me understand a couple of things.

I should have specified the issue is when using Rust calls Dart which must uses DCO by default?

```rust
//dart_minimal/rust/src/api/minimal.rs

//added code
pub struct SomeType {
    pub field1: i64,
}
//this works
pub fn takes_some_type(var: SomeType) {
    println!("got var {}",var.field1);
}
//this will cause the issue
pub async fn takes_a_cb(cb: impl Fn(SomeType) -> DartFnFuture<()> + Send + Sync + 'static) {
    cb(SomeType { field1: 200 }).await
}
```

```dart
//dart_minimal/lib/main.dart

//added code

//fails with "UnimplementedError: Not implemented in this codec, please use the other one"
takesACb(cb: (SomeType ty) => {print("got sometype ${ty.field1}")}); 

//works becuase of sse
SomeType val = SomeType(field1: 100);
takesSomeType(var_: val);
```
So the impl is just missing...I'll try to put together a PR for this scenario

**Reply by sampaioletti** [reply]
adding this match arm in frb_codegen\src\library\codegen\generator\wire\dart\spec_generator\codec\dco\decoder\ty\delegate.rs seems to work well in my project. I'll have to get more familiar with the project so I can submit a pr with tests etc.

```rust
MirTypeDelegate::CastedPrimitive(mir) => {
    let postfix = match mir.inner {
        MirTypePrimitive::I64  => "i_64",
        MirTypePrimitive::Isize  => "isize",
        MirTypePrimitive::U64 => "u_64",
        MirTypePrimitive::Usize  => "usize",
        // frb-coverage:ignore-start
        _ => return "throw UnimplementedError('Not implemented in this codec, please use the other one');".into(),
        // frb-coverage:ignore-end
    };
    format!(r"return dco_decode_{}(raw);",postfix)
}
```

**Reply by fzyzcjy** [reply]
Hmm, yes callbacks' default codec may be different, though for consistency it may should use sse by default... but dco is reasonable

For tests: If you want to test such rust-callback-to-dart things, then tests are in https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/frb_example/pure_dart/rust/src/api/dart_fn.rs
